zkvm/src/program.rs

```diff
@@ -120,56 +131,69 @@ impl Program {
         self
     }
 
-    /// Takes predicate and closure to add choose operations for
-    /// predicate tree traversal.
-    pub fn choose_predicate<F, T>(
+    /// Takes predicate tree and index of program in Merkle tree to verify
+    /// the program's membership in that Merkle tree and call the program.
+    pub fn choose_call(
         &mut self,
-        pred: Predicate,
-        choose_fn: F,
-    ) -> Result<&mut Program, VMError>
-    where
-        F: FnOnce(PredicateTree) -> Result<T, VMError>,
-    {
-        choose_fn(PredicateTree {
-            prog: self,
-            pred: pred,
-        })?;
+        pred_tree: PredicateTree,
+        prog_index: usize,
+    ) -> Result<&mut Program, VMError> {
+        let (call_proof, program) = pred_tree.create_callproof_program(prog_index)?;
+        self.push(Data::Opaque(call_proof.to_bytes())).call();
+        self.push(Data::Program(program)).call();
         Ok(self)
     }
 }
 
-/// Adds data and instructions to traverse a predicate tree.
-pub struct PredicateTree<'a> {
-    prog: &'a mut Program,
-    pred: Predicate,
-}
+impl ProgramItem {
+    /// Returns the number of bytes needed to serialize the ProgramItem.
+    pub fn serialized_length(&self) -> usize {
+        match self {
+            ProgramItem::Program(prog) => prog.serialized_length(),
+            ProgramItem::Bytecode(vec) => vec.len(),
+        }
+    }
 
-impl<'a> PredicateTree<'a> {
-    /// Kth Predicate branch
-    pub fn select(self, k: usize) -> Result<Self, VMError> {
-        let preds = self.pred.to_disjunction()?;
-        let n = preds.len();
-        let selected = preds[k].clone();
-        if k >= n {
-            return Err(VMError::PredicateIndexInvalid);
+    /// Encodes a program item into a buffer.
+    pub fn encode(&self, buf: &mut Vec<u8>) {
+        match self {
+            ProgramItem::Program(prog) => prog.encode(buf),
+            ProgramItem::Bytecode(bytes) => {
+                buf.extend_from_slice(&bytes);
+            }
         }
-        let prog = self.prog;
-        for pred in preds.iter() {
-            prog.push(pred.as_opaque());
+    }
+
+    /// Downcasts a program item into a program.
+    pub fn to_program(self) -> Result<Program, VMError> {
+        match self {
+            ProgramItem::Program(prog) => Ok(prog),
+            ProgramItem::Bytecode(_) => return Err(VMError::TypeNotProgram),
+        }
+    }
+
+    /// Downcasts a program item into a vector of bytes.
+    pub fn to_bytes(self) -> Result<Vec<u8>, VMError> {
+        match self {
+            ProgramItem::Program(_) => return Err(VMError::TypeNotProgram),
```

Comment:
If you are sure we don't need this method to force-covert non-opaque program to bytes, then it's better to rename it to something sounding more like "downcast", not encoding. E.g. `to_bytecode`. And clarify that the ::Program is not encoded automatically - for this one must use `encode`.